src/rpc/blockchain.cpp

```diff
@@ -1240,18 +1232,20 @@ static UniValue BIP9SoftForkDesc(const Consensus::Params& consensusParams, Conse
         statsUV.pushKV("elapsed", statsStruct.elapsed);
         statsUV.pushKV("count", statsStruct.count);
         statsUV.pushKV("possible", statsStruct.possible);
-        rv.pushKV("statistics", statsUV);
+        bip9.pushKV("statistics", statsUV);
     }
-    return rv;
-}
 
-static void BIP9SoftForkDescPushBack(UniValue& bip9_softforks, const Consensus::Params& consensusParams, Consensus::DeploymentPos id)
-{
-    // Deployments with timeout value of 0 are hidden.
-    // A timeout value of 0 guarantees a softfork will never be activated.
-    // This is used when softfork codes are merged without specifying the deployment schedule.
-    if (consensusParams.vDeployments[id].nTimeout > 0)
-        bip9_softforks.pushKV(VersionBitsDeploymentInfo[id].name, BIP9SoftForkDesc(consensusParams, id));
+    UniValue rv(UniValue::VOBJ);
+    rv.pushKV("type", "bip9");
+    rv.pushKV("bip9", bip9);
+    if (ThresholdState::LOCKED_IN == thresholdState) {
+        rv.pushKV("height", since_height + consensusParams.nMinerConfirmationWindow);
```

Comment:
I don't love that we're leaking bip9 implementation calculations into the rpc code.  Though it is useful to see when a LOCKED_IN softfork will actually activate.  Can we invoke a method in the bip9 machinery instead?

Reply:
Done. Added a `VersionBitsActivationHeight()` function to validation.cpp